### PR TITLE
#1062

### PR DESCRIPTION
**Pull request by Nikita-baghela07**
Add contextual typing implementation and tests

Implements contextual type inference for Teal, enabling the compiler to infer types from context rather than requiring explicit annotations. Includes:

Core contextual typing engine with type inference optimization
Generic constraint satisfaction and resolution
Mixed parameter handling with improved error reporting
Comprehensive test suite covering basic inference, backward compatibility, and performance
Documentation on contextual typing usage and troubleshooting
Parser extensions to support contextual type expressions
This enhancement improves developer experience by reducing boilerplate type annotations while maintaining full type safety.

### CODE REVIEW

**Comment by mtdowling** [comment]
This is really cool, and I would love this feature. Do you think it would be possible to break this into smaller PRs, or is this all tightly interconnected? e.g., it looks like this includes the duplicate interface fix and the diamond dependency fix. Are those pre-reqs?

File: spec/lang/subtyping/interface_spec.lua [hunk 1]
```diff
@@ -257,38 +257,31 @@ describe("subtyping of interfaces:", function()
...
+
+      -- Test that all fields are accessible
+      local d: D = {
+         a = 1,    -- from A (via B and C)
+         b = "hi", -- from B
+         c = true, -- from C
+         d = 42    -- from D
+      }
+      print(d.a, d.b, d.c, d.d)
+   ]]))
```
Comment (mbartlett21) [inline spec/lang/subtyping/interface_spec.lua]: It is impossible to test the duplicate interface list since that is an internal compiler detail. This test already passes on main:

[clicky](https://teal-playground.netlify.app/?c=CiAgICAgIGxvY2FsIGludGVyZmFjZSBBCiAgICAgICAgIGE6IG51bWJlcgogICAgICBlbmQKCiAgICAgIGxvY2FsIGludGVyZmFjZSBCIGlzIEEKICAgICAgICAgYjogc3RyaW5nCiAgICAgIGVuZAoKICAgICAgbG9jYWwgaW50ZXJmYWNlIEMgaXMgQQogICAgICAgICBjOiBib29sZWFuCiAgICAgIGVuZAoKICAgICAgLS0gVGhpcyBzaG91bGQgcmVzdWx0IGluIGludGVyZmFjZV9saXN0OiB7QiwgQSwgQ30gKG5vIGR1cGxpY2F0ZSBBKQogICAgICBsb2NhbCByZWNvcmQgRCBpcyBCLCBDCiAgICAgICAgIGQ6IGludGVnZXIKICAgICAgZW5kCgogICAgICAtLSBUZXN0IHRoYXQgYWxsIGZpZWxkcyBhcmUgYWNjZXNzaWJsZQogICAgICBsb2NhbCBkOiBEID0gewogICAgICAgICBhID0gMSwgICAgLS0gZnJvbSBBICh2aWEgQiBhbmQgQykKICAgICAgICAgYiA9ICJoaSIsIC0tIGZyb20gQgogICAgICAgICBjID0gdHJ1ZSwgLS0gZnJvbSBDCiAgICAgICAgIGQgPSA0MiAgICAtLSBmcm9tIEQKICAgICAgfQogICAgICBwcmludChkLmEsIGQuYiwgZC5jLCBkLmQpCg%3D%3D)

File: spec/lang/contextual_typing/backward_compatibility_property_spec.lua [hunk 2]
```diff
@@ -0,0 +1,530 @@
+local util = require("spec.util")
+local tl = require("teal.api.v2")
+local contextual_typing = require("teal.contextual_typing")
+local contextual_type_checker = require("teal.contextual_type_checker")
+
+describe("contextual typing backward compatibility property", function()
+   print("DEBUG: Starting backward compatibility property tests")
+   
+   -- Property 4: Backward Compatibility
+   -- **Validates: Requirements 1.4, 5.1, 5.4**
```
Comment (mbartlett21) [inline spec/lang/contextual_typing/backward_compatibility_property_spec.lua]: What do these numbers refer to? I find no reference apart from these comments.

File: spec/lang/contextual_typing/generic_constraint_satisfaction_spec.lua [hunk 3]
```diff
@@ -0,0 +1,795 @@
...
+local tl = require("teal.api.v2")
+local contextual_typing = require("teal.contextual_typing")
+local generic_resolution = require("teal.generic_resolution")
+local generic_resolver = require("teal.generic_resolver")
+
+describe("contextual typing", function()
+   describe("generic constraint satisfaction property", function()
+      
+      -- Property 5: Generic Constraint Satisfaction
+      -- **Validates: Requirements 2.1, 2.2, 8.1**
```
Comment (mbartlett21) [inline spec/lang/contextual_typing/generic_constraint_satisfaction_spec.lua]: Same comment as above with unreferenced numbers.

File: spec/lang/contextual_typing/backward_compatibility_property_spec.lua [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+      -- Compare AST structures
+      local compatible, errors = checker:compare_ast_structures(ast1, ast2)
+      
+      assert.is_not_nil(compatible, "Should get comparison result")
+      assert.is_not_nil(errors, "Should get errors array")
+   end)
+   
+   -- Test edge cases
+   it("validates backward compatibility handles edge cases", function()
+      local checker = contextual_type_checker.BaseContextualTypeChecker:new()
```
Comment (mbartlett21) [inline spec/lang/contextual_typing/backward_compatibility_property_spec.lua]: This is unused

File: spec/lang/contextual_typing/backward_compatibility_property_spec.lua [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+      
+      -- Empty function
+      local empty_code = "local function f() end"
+      local ast1, errors1 = tl.parse(empty_code, "test.tl")
+      assert.equal(0, #errors1, "Should parse empty function")
+      
+      -- Function with many parameters
+      local many_params = "local function f(a: number, b: number, c: number, d: number, e: number): number return a + b + c + d + e end"
```
Comment (mbartlett21) [inline spec/lang/contextual_typing/backward_compatibility_property_spec.lua]: This is not many parameters.
You would probably want something close to the Lua limit (I think ~200?).

File: spec/lang/contextual_typing/parser_extensions_spec.lua [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+describe("contextual typing parser extensions", function()
+   
+   describe("untyped parameter parsing", function()
+      it("parses function with single untyped parameter", util.check([[
+         local f = function(x) return x end
+      ]]))
+      
+      it("parses function with multiple untyped parameters", util.check([[
+         local f = function(a, b, c) return a + b + c end
```
Comment (mbartlett21) [inline spec/lang/contextual_typing/parser_extensions_spec.lua]: I don't think we want to allow this with unresolved types.

File: spec/lang/contextual_typing/backward_compatibility_property_spec.lua [hunk 7]
```diff
@@ -0,0 +1,530 @@
+local util = require("spec.util")
+local tl = require("teal.api.v2")
+local contextual_typing = require("teal.contextual_typing")
+local contextual_type_checker = require("teal.contextual_type_checker")
+
+describe("contextual typing backward compatibility property", function()
+   print("DEBUG: Starting backward compatibility property tests")
+
```
Comment (mbartlett21) [inline spec/lang/contextual_typing/backward_compatibility_property_spec.lua]: Please don't have trailing spaces or unnecessary spaces.

(The empty line is fine, it is just that it is filled with three space characters as well)

`grep -rl '  *$' ./*` will find all the files with trailing spaces.

And then `sed -i 's/  *$//' FILE...` will replace the spaces with true emptiness.

**Comment by Nikita-baghela07** [comment]
No, it's not wholly AI-generated, but I took some help for some parts, and it's causing me this.
i'll try to implement it again with another method and in small PRs.
